packages/dashboard-message-bus/lib/DashboardMessageBus.ts

```diff
@@ -0,0 +1,178 @@
+
+  constructor(
+    public requestsPort: number,
+    public listenPort: number,
+    public host: string = "0.0.0.0"
+  ) {
+    super();
+  }
+
+  start() {
+    this.listenServer = new WebSocket.Server({
+      host: this.host,
+      port: this.listenPort
+    });
+
+    this.listenServer.on("connection", (newListener: WebSocket) => {
+      newListener.on("close", () => {
+        this.logToClients("Listener disconnected", "connections");
+
+        this.listeningSockets = this.listeningSockets.filter(
+          listener => listener !== newListener
+        );
+        this.terminateIfNoConnections();
+      });
+
+      // Process all backlogged (unfulfilled) requests on new listener connection.
+      this.unfulfilledRequests.forEach(({ socket, data }) =>
+        this.processRequest(socket, data, [newListener])
+      );
+
+      this.logToClients("Listener connected", "connections");
+
+      this.listeningSockets.push(newListener);
+    });
+
+    this.requestsServer = new WebSocket.Server({
+      host: this.host,
+      port: this.requestsPort
+    });
+
+    this.requestsServer.on("connection", (newClient: WebSocket) => {
+      newClient.on("message", (data: WebSocket.Data) => {
+        this.processRequest(newClient, data, this.listeningSockets);
+      });
+
+      newClient.on("close", () => {
+        this.logToClients("Client disconnected", "connections");
+
+        this.clientSockets = this.clientSockets.filter(
+          client => client !== newClient
+        );
+        this.clearClientRequests(newClient);
+        this.terminateIfNoConnections();
+      });
+
+      this.logToClients("Client connected", "connections");
+
+      this.clientSockets.push(newClient);
+    });
+  }
+
+  async ready() {
+    if (this.listeningSockets.length > 0) return;
+    await delay(1000);
+    await this.ready();
+  }
+
+  async terminate() {
+    await promisify(this.requestsServer.close.bind(this.requestsServer))();
+    await promisify(this.listenServer.close.bind(this.listenServer))();
+    this.emit("terminate");
+  }
+
+  private async processRequest(
+    socket: WebSocket,
+    data: WebSocket.Data,
+    listeners: WebSocket[]
+  ) {
+    if (typeof data !== "string") return;
+    await this.ready();
+
+    this.unfulfilledRequests.set(data, { socket, data });
+    const message = base64ToJson(data);
+
+    try {
+      this.logToClients(
+        `Sending message to ${listeners.length} listeners`,
+        "requests"
+      );
+      this.logToClients(message, "requests");
+
+      const response = await broadcastAndAwaitFirst(listeners, message);
+
+      this.logToClients(
+        `Sending response for message ${message.id}`,
+        "responses"
+      );
+      this.logToClients(response, "responses");
+
+      const encodedResponse = jsonToBase64(response);
+      socket.send(encodedResponse);
+      this.unfulfilledRequests.delete(data);
+
+      this.invalidateMessage(message.id);
```

Comment:
❓ Why do we send "invalidate" messages once requests have been processed?  Is this to prevent another instance from processing it or something?

Reply:
Yeah that's it. If you have two dashboard windows open then all requests get sent to both dashboards. After processing the request in one dashboard, they would persist in the other dashboard. This "invalidate" step tells that other dashboard that the request no longer needs to be fulfilled.